src/dmd/backend/cod3.d

```diff
@@ -2595,21 +2595,74 @@ code *genmovreg(uint to,uint from)
 
 void genmovreg(ref CodeBuilder cdb,uint to,uint from)
 {
-    debug
-        if (to > ES || from > ES)
-            printf("genmovreg(to = %d, from = %d)\n",to,from);
+    genmovreg(cdb, to, from, TYMAX);
+}
+
+void genmovreg(ref CodeBuilder cdb, uint to, uint from, tym_t tym)
+{
+    // register kind. ex: GPR,XMM,SEG
+    static uint _K(uint reg)
+    {
+        switch (reg)
+        {
+        case ES:                   return ES;
+        case XMM15:
+        case XMM0: .. case XMM7:   return XMM0;
+        case AX:   .. case R15:    return AX;
+        default:                   return reg;
+        }
+    }
+
+    // kind combination (order kept)
+    static uint _X(uint to, uint from) { return (_K(to) << 8) + _K(from); }
 
-    assert(to <= ES && from <= ES);
     if (to != from)
     {
-        if (to == ES)
-            genregs(cdb,0x8E,0,from);
-        else if (from == ES)
-            genregs(cdb,0x8C,0,to);
-        else
-            genregs(cdb,0x89,from,to);
-        if (I64)
-            code_orrex(cdb.last(), REX_W);
+        if (tym == TYMAX) tym = TYsize_t; // avoid register slicing
```

Comment:
??? register slicing ???

Reply:
Line `2598` uses `TYMAX` as a placeholder type.  `TYMAX` has no size, so if no size specified, this moves a full register.